File: Python-version/human.py
from animal import Animal

class Human(Animal):
    def __init__(self, x, y, world, oid=0, s=None, i=None, a=None, m=None):
        super().__init__(x, y, world, oid, str, i, a, m)
        self.sign='Y'
        self.name="Human"
        self.color="#FF88FD"
        self.str=5
        self.ini=4
        self.speciesId=11
        self.cooldown=0
        self.power_left=0
        self.action=''
        self.change_x=0
        self.change_y = 0
        self.special_active=False
        self.moved=False

    def MakeNewA(self,x,y):
        return None

    def GetCd(self):
        return self.cooldown

    def SetCd(self,cd):
        self.cooldown=cd

    def GetP(self):
        return self.power_left

    def SetP(self,p):
        self.power_left=p

    def SetAction(self,ac):
        self.action=ac

    def TakeAction(self):

        if self.cooldown==10:
            print(f"Alzur's shield activated")
            self.special_active = True

        if self.action=='w':
            self.change_y-=1
        elif self.action=='s':
            self.change_y+=1
        elif self.action=='a':
            self.change_x-=1
        elif self.action=='d':
            self.change_x+=1

        if self.xPos + self.change_x < self.world.GetW() and self.yPos + self.change_y < self.world.GetH() and self.xPos + self.change_x >= 0 and self.yPos + self.change_y >= 0:
            potential_collision = self.world.LookForCollision(self.xPos + self.change_x, self.yPos + self.change_y)

            if potential_collision is not None:
                self.Collision(potential_collision)

            if self.age != -1:
                potential_collision = self.world.LookForCollision(self.xPos + self.change_x, self.yPos + self.change_y)
                if potential_collision is None:
                    tmpx = self.xPos + self.change_x
                    tmpy = self.yPos + self.change_y
                    print(f"{self.name} moved from tile ({self.xPos}, {self.yPos}) to tile ({tmpx}, {tmpy})")
                    self.log = f"{self.name} moved from tile ({self.xPos}, {self.yPos}) to tile ({tmpx}, {tmpy})"
                    self.xPos += self.change_x
                    self.yPos += self.change_y

            self.moved=True
            self.world.AddLog(self.log)
            self.log = " "
            if self.age > 0:
                self.age += 1
            if self.cooldown>0:
                self.cooldown-=1
            if self.power_left>0:
                self.power_left-=1
            self.change_x=0
            self.change_y=0
            if self.power_left==0:
                self.special_active=False

    def Win(self,othercreature):
        if not self.special_active:#zabezpieczenie, bo jak mamy aktywna tarcze to nie wygramy per se tylko odepchniemy
            print(f" and {self.name} won! {othercreature.GetName()} perished")
            self.world.RmOrganism(othercreature.GetOId())
            othercreature.ResetLog()
            othercreature.SetAge(-1)

    def Defended(self,attackingcreature):
        if self.special_active:
            attackingcreature.TakeAction()
            print(f" and {self.name} won! {attackingcreature.GetName()} was pushed back")
            return True
        if self.str>attackingcreature.GetS():
            return True
        return False

    def print(self):
        self.world.SetMap(self.xPos,self.yPos,self.sign,self.color)

File: Java-version/Human.java
import java.awt.*;
import java.util.Random;

public class Human extends Animal{
    public Human(int x, int y, World world)
    {super(x,y,world);
        sign = 'Y';
        name = "Human";
        color = Color.BLACK;
        str = 5;
        ini = 4;
        speciesId = 11;
    }
    public Human(int str, int ini, int xPos, int yPos, int age, int oId, boolean moved, World world)
    {super(str, ini, xPos, yPos, age, oId, moved, world);
        sign = 'Y';
        name = "Human";
        speciesId = 11;
        color = Color.BLACK;
    }
    @Override
    public Animal MakeNewA(int x, int y) {
        return null;
    }

    public  int GetCd() {
        return cooldown;
    }
    public void SetCd(int cd) {
        cooldown = cd;
    }
    public  int GetP() {
        return powerLeft;
    }
    public void SetP(int p) {
        powerLeft = p;
    }

    public void SetAction(char ac) {
        action = ac;
    }

    public void TakeAction() {
        if (powerLeft == 5) {
                specialActivated = true;
                System.out.print("Azure's shield activated!" + '\n');
                log = log + "Azure's shield activated!" + '\n';
        }

        if (action == 'w')
            changeY--;
        else if (action == 's')
            changeY++;
        else if (action == 'a')
            changeX--;
        else if (action == 'd')
            changeX++;


        if (xPos + changeX < world.GetW() && yPos + changeY < world.GetH() &&  xPos + changeX >= 0 && yPos + changeY >= 0) {
            Organism potentialCollision = world.LookForCollision(xPos + changeX, yPos + changeY);
            if (potentialCollision != null) {
                Collision(potentialCollision);
            }
            if (age != DEAD) {
                potentialCollision = world.LookForCollision(xPos + changeX, yPos + changeY);
                if (potentialCollision == null) {
                    int tmpx = xPos + changeX;
                    int tmpy = yPos + changeY;
                    System.out.print(name + " moved from tile (" + xPos + ", " + yPos + ") to tile (" + tmpx + ", " + tmpy + ")" + '\n');
                    log = log + name + " moved from tile (" + xPos + ", " + yPos + ") to tile (" + tmpx + ", " + tmpy + ")" + '\n';
                    xPos += changeX;
                    yPos += changeY;
                }
            }
        }
        world.AddLog(log);
        log = " ";
        moved = true;
        if (age > 0) age++;
        if (cooldown > 0)
            cooldown--;
        if (powerLeft > 0)
            powerLeft--;
        changeX = 0;
        changeY = 0;
        if (powerLeft==0){
        specialActivated = false;
        }
    }
    @Override
    public void Win(Organism otherCreature){
        if (!specialActivated){
            System.out.print(" and" + name + " won!" + otherCreature.name + " perished" + '\n');
            log= log + " and" + name + " won!" + otherCreature.name +  " perished" + '\n';
            world.RmOrganism(otherCreature.oId);

            world.AddLog(otherCreature.GetLog());
            otherCreature.ResetLog();
            otherCreature.SetAge(DEAD);
        }
    }
    @Override
    public boolean Defended(Organism attackingCreature) {
        if (specialActivated){
            attackingCreature.TakeAction();
            System.out.print(" and" + name + " won!" + attackingCreature.name + " was pushed back" + '\n');
            log= log + " and" + name + " won!" + attackingCreature.name + " was pushed back" + '\n';
            return true;
        }
        else if (str>attackingCreature.GetS()){
            return true;
        }
        return false;
    }

    private int cooldown = 0, powerLeft = 0;
    private char action;
    private int changeX = 0, changeY = 0;
    private boolean specialActivated = false;
}
